gcc/rust/expand/rust-macro-builtins-asm.cc

```diff
@@ -16,5 +16,528 @@
+    }
+  else
+    {
+      return tl::nullopt;
+    }
+  return reg_operand;
+}
+void
+check_and_set (Parser<MacroInvocLexer> &parser, InlineAsmContext &inlineAsmCtx,
+	       AST::InlineAsmOption option)
+{
+  auto &inlineAsm = inlineAsmCtx.inlineAsm;
+  if (inlineAsm.options.count (option) != 0)
+    {
+      // TODO: report an error of duplication
+      rust_error_at (parser.peek_current_token ()->get_locus (),
+		     "the `%s` option was already provided",
+		     InlineAsmOptionMap[option].c_str ());
+      return;
+    }
+  else
+    {
+      inlineAsm.options.insert (option);
+    }
+}
+int
+parse_options (Parser<MacroInvocLexer> &parser, TokenId last_token_id,
+	       InlineAsmContext &inlineAsmCtx)
+{
+  bool is_global_asm = inlineAsmCtx.inlineAsm.is_global_asm;
+  // Parse everything commitedly
+  if (!parser.skip_token (LEFT_PAREN))
+    {
+      // We have shifted `options` to search for the left parenthesis next, we
+      // should error out if this is not possible.
+      // TODO: report some error.
+      return -1;
+    }
+
+  auto token = parser.peek_current_token ();
+  while (token->get_id () != last_token_id && token->get_id () != RIGHT_PAREN)
+    {
+      if (!is_global_asm && check_identifier (parser, "pure"))
+	{
+	  check_and_set (parser, inlineAsmCtx, AST::InlineAsmOption::PURE);
+	}
+      else if (!is_global_asm && check_identifier (parser, "nomem"))
+	{
+	  check_and_set (parser, inlineAsmCtx, AST::InlineAsmOption::NOMEM);
+	}
+      else if (!is_global_asm && check_identifier (parser, "readonly"))
+	{
+	  check_and_set (parser, inlineAsmCtx, AST::InlineAsmOption::READONLY);
+	}
+      else if (!is_global_asm && check_identifier (parser, "preserves_flags"))
+	{
+	  check_and_set (parser, inlineAsmCtx,
+			 AST::InlineAsmOption::PRESERVES_FLAGS);
+	}
+      else if (!is_global_asm && check_identifier (parser, "noreturn"))
+	{
+	  check_and_set (parser, inlineAsmCtx, AST::InlineAsmOption::NORETURN);
+	}
+      else if (!is_global_asm && check_identifier (parser, "nostack"))
+	{
+	  check_and_set (parser, inlineAsmCtx, AST::InlineAsmOption::NOSTACK);
+	}
+      else if (!is_global_asm && check_identifier (parser, "may_unwind"))
+	{
+	  check_and_set (parser, inlineAsmCtx,
+			 AST::InlineAsmOption::MAY_UNWIND);
+	}
+      else if (check_identifier (parser, "att_syntax"))
+	{
+	  check_and_set (parser, inlineAsmCtx,
+			 AST::InlineAsmOption::ATT_SYNTAX);
+	}
+      else if (check_identifier (parser, "raw"))
+	{
+	  check_and_set (parser, inlineAsmCtx, AST::InlineAsmOption::RAW);
+	}
+      else
+	{
+	  // TODO: Unexpected error, please return the correct error
+	  rust_error_at (token->get_locus (),
+			 "Unexpected token encountered in parse_options");
+	}
+      if (parser.skip_token (RIGHT_PAREN))
+	{
+	  break;
+	}
+
+      // Parse comma as optional
+      if (parser.skip_token (COMMA))
+	{
+	  continue;
+	}
+      else
+	{
+	  std::cout << "Sum wrong here, check it out" << std::endl;
```

Comment:
:smile: 

I think we can keep that in a `rust_debug` but it should be removed from the PR :P

Reply:
💀💀💀💀 i have no idea how i let this slipped through the post commit stage 💀💀💀